DataDog/graft-rb
# Add SimpleCov reporting with GitHub Job Summary

### Why?


Providing code coverage in a simple way.

### What does this PR do?


- [X] Introduce SimpleCov as a `spec_helper` plugin
- [X] Generate human-oriented HTML coverage report
- [X] Generate machine-oriented JSON coverage report
- [X] Add support for a `COVERAGE` environment variable
- [X] Add a Markdown formatter for SimpleCov
- [X] Use Markdown output as a [GitHub Job Summary](https://github.blog/2022-05-09-supercharging-github-actions-with-job-summaries/)
- [ ] Archive HTML coverage as a downloadable job artifact
- [ ] Improve Markdown output readability when unrendered (e.g table alignment)
- [ ] Improve Markdown output to generate full annotated source output / "worst offender" summary for each source file
- [ ] [Stretch] Integrate coverage with editors and IDEs (e.g VSCode)
- [ ] [Stretch] Extract Markdown formatter to its own gem

### How to test the change?


- Run bundle exec rake spec
- Observe HTML, JSON, and Markdown output in `coverage`

### Additional Notes:


The Markdown formatter is based on the SimpleCov HTML reporter, essentially inheriting from it, overriding a few HTML rendering functions, and defining new views. Reason is that if better formatting is desired then Markdown can easily embed HTML (at the cost of making the pure markdown output less readable.

There are other Markdown formatters out there but they are unmaintained, so I chose to vendor a fresh new piece of code.
